Test sharding

VenelinMartinov opened the pull request:
This PR implements a test sharding strategy to speed up the bridge tests from 1 hour 30 minutes to under 15 minutes.

It uses https://github.com/VenelinMartinov/sharder which is a package for generating regexes to plug into `go test` based on the test profile data. The script is based on the previous work by @blampe in https://github.com/pulumi/shard.

This PR also adds two new `make` commands:
- `make test_profile` this regenerates the test profile data, so that shard partitioning can be more fair
- `make test_shard`. This is only run in CI. This command picks up the `SHARD_CMD` env var to determine which tests to run. `SHARD_CMD` is populated in CI by the `sharder` package and contains a regex which should match the correct tests to run.

Sharding is only done for `ubuntu` tests as they take the longest.

Example run: https://github.com/pulumi/pulumi-terraform-bridge/actions/runs/12656374664

iwahbe commented on Makefile:
```diff
@@ -33,6 +33,20 @@ test:: install_plugins
+	PULUMI_TERRAFORM_BRIDGE_TEST_PROVIDER=$(shell pwd)/bin/pulumi-terraform-bridge-test-provider \
+		go test -count=1 -coverprofile="coverage.txt" -coverpkg=./... -timeout 2h -parallel ${TESTPARALLELISM} ./... ${SHARD_CMD}
+
+test_profile:: install_plugins
```
This needs a comment explaining what it does.

iwahbe commented on Makefile:
```diff
@@ -33,6 +33,20 @@ test:: install_plugins
 	PULUMI_TERRAFORM_BRIDGE_TEST_PROVIDER=$(shell pwd)/bin/pulumi-terraform-bridge-test-provider \
 		go test -count=1 -coverprofile="coverage.txt" -coverpkg=./... -timeout 2h -parallel ${TESTPARALLELISM} ./...
 
+test_shard:: install_plugins
```
This also needs a comment explaining what it does, and that it is provided for CI.

iwahbe commented on .gitignore:
```diff
@@ -14,4 +14,6 @@ testing/coverage.txt
\ No newline at end of file
+pulumi-hcl-lint
+
+schema-tpsdkv2.json
```
Why are we adding this. We should fix the test that erroneously generates this file instead of ignoring it.

iwahbe commented on .github/workflows/build-and-test.yml:
```diff
@@ -21,6 +22,68 @@ jobs:
+            feature-flags: "PULUMI_TF_BRIDGE_ACCURATE_BRIDGE_PREVIEW"
+          - platform: macos-latest
+            feature-flags: "PULUMI_TF_BRIDGE_ACCURATE_BRIDGE_PREVIEW"
+          - platform: windows-latest
+            shard: 1
```
Is the idea that every OS runs shard 0, and only ubuntu runs the rest of the shards? If so, that needs a comment explaining so.

Also, why don't we shard all OSs?

VenelinMartinov replied:
Mac and windows test are fairly quick so we don't need to shard them.

iwahbe commented on .github/workflows/build-and-test.yml:
```diff
@@ -21,6 +22,68 @@ jobs:
+            feature-flags: "PULUMI_TF_BRIDGE_ACCURATE_BRIDGE_PREVIEW"
+          - platform: macos-latest
+            feature-flags: "PULUMI_TF_BRIDGE_ACCURATE_BRIDGE_PREVIEW"
+          - platform: windows-latest
```
Instead of defining `shard` as `[0, 1, 2, 3, ...]`, then excluding each option we don't want, can we define `shard` as `[0]` and then add explicit includes for other shard options. That will reduce the length of the shard list by half.

iwahbe commented on .github/workflows/build-and-test.yml:
```diff
@@ -35,18 +98,25 @@ jobs:
       - name: Build PF
         run: cd pkg/pf && make build
+      - name: Shard tests
+        run: echo "SHARD_CMD=$(go run github.com/VenelinMartinov/sharder@9d09afeb1c053b4a0263fbaa5c3c9da1ca2d10e7 --total ${{ env.TOTAL_SHARDS }} --index ${{ matrix.shard }} --output testoutput.txt --format make)" >> $GITHUB_ENV
```
How is `github.com/VenelinMartinov/sharder` different from `github.com/pulumi/shard`? Why can't we use `github.com/pulumi/shard`?

VenelinMartinov replied:
The two are completely separate modules. `pulumi/shard` parses the test files to find the tests and then partitions them randomly.

`sharder` uses test profiling data to attempt a fair paritition. I only added that since `shard` did little for the bridge.

t0yv0 replied:
Perhaps merge this functionality into pulumi/shard?

VenelinMartinov replied:
I considered it but the two share little except the external interface. Seems less complicated to keep them separate.

t0yv0 replied:
So you decided to add a dependency on a private repo instead? I think this will raise some eyebrows. Perhaps compute a config using your script and pass it to pulumi/shard? Or discuss with the team?

VenelinMartinov replied:
I'm more than happy to move the repo under pulumi but that would have slowed down the iteration here. I wanted to get this out of the door as it is literally going to save us hours of waiting.

VenelinMartinov replied:
Happy to take other suggestions here too.

VenelinMartinov replied:
I've moved this to use a repo under `pulumi/`

iwahbe reviewed (commented):
If I add new tests, do I need to re-run `make test_profile`? If so, that means that I need to re-run the entire test suite on my laptop to add a single test (which is a major usability regression). If not, then I don't understand how the test profile gets updated. 

github.com/pulumi/sharder's REAMDE is very sparse, which doesn't give me confidence. I'd like to be able to read the README and understand how the tool works, not just a short explanation of what it does. The README should also explain the irreconcilable differences between github.com/pulumi/sharder and github.com/pulumi/shard.

VenelinMartinov commented:
@iwahbe I've updated the README on `sharder`. Let me know if that is not clearer.

>If I add new tests, do I need to re-run make test_profile?

No, that is not necessary. I've added a description to the `sharder` README about it. The TLDR is that the last shard will run all tests not run by any previous ones. This means that if the test timings are not updated the shards might become unbalanced but no tests will be skipped.

In practice, I think this will mean that we'll re-record them once we add a good chunk of tests but don't need to touch that otherwise.

iwahbe commented:
Closing in favor of https://github.com/pulumi/pulumi-terraform-bridge/pull/2832, which was stacked on top of this PR.
